### PR TITLE
Core/SAI: fix SMART_EVENT_SPELLHIT with gobs as caster

### LINKED ISSUE #22613: Core/SAI: SMART_EVENT_SPELLHIT with gobs as caster
**Description:** The event SMART_EVENT_SPELLHIT will not be triggered if the unit that cast the spell is a game Object.

``` sql
-- Servant of Razelikh
UPDATE `creature_template` SET `AIName`="SmartAI" WHERE `entry`=7668;
DELETE FROM `smart_scripts` WHERE `entryorguid`=7668 AND `source_type`=0;
DELETE FROM `smart_scripts` WHERE `entryorguid`=766800 AND `source_type`=9;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(7668,0,0,0,25,0,100,0,0,0,0,0,42,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Servant of Razelikh - On Reset - Set Invincibility at 1 HP"),
(7668,0,1,0,4,0,100,0,0,0,0,0,64,1,0,0,0,0,0,7,0,0,0,0,0,0,0,"Servant of Razelikh - On Aggro - Store Target"),
(7668,0,2,0,2,0,100,1,0,5,0,0,11,10794,0,0,0,0,0,1,0,0,0,0,0,0,0,"Servant of Razelikh - Between 0-5% Health - Cast 'Spirit Shock' (No Repeat)"),
(7668,0,3,0,2,0,100,1,0,5,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Servant of Razelikh - Between 0-5% Health - Say Line 0 (No Repeat)"),
(7668,0,4,0,2,0,100,1,0,5,0,0,106,16,0,0,0,0,0,15,141812,10,0,0,0,0,0,"Servant of Razelikh - Between 0-5% Health - Remove Gameobject Flag 'Not Selectable' (Stone of Binding)"),
(7668,0,5,0,8,0,100,1,10805,0,0,0,80,766800,2,0,0,0,0,1,0,0,0,0,0,0,0,"Servant of Razelikh - On Spellhit 'Shackle Shatter' - Run Script"),
(766800,9,0,0,0,0,100,0,3000,3000,0,0,1,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Servant of Razelikh - On Spellhit 'Shackle Shatter' - Say Line 1"),
(766800,9,1,0,0,0,100,0,0,0,0,0,33,7668,0,0,0,0,0,12,1,0,0,0,0,0,0,"Servant of Razelikh - On Spellhit 'Shackle Shatter' - Quest Credit 'The Stones That Bind Us'"),
(766800,9,2,0,0,0,100,0,0,0,0,0,37,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Servant of Razelikh - On Spellhit 'Shackle Shatter' - Kill Self");

DELETE FROM `creature_text` WHERE `CreatureID`=7668;
INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
(7668,0,0,"%s's physical form is weakened. You notice the Stone of Binding near the creature is glowing! Strike now!",41,0,100,0,0,0,3642,0,"Servant of Razelikh"),
(7668,1,0,"FREEEEEEEEEEDDOOOOMM!",12,0,100,0,0,0,3641,0,"Servant of Razelikh"),
(7668,1,1,"Finally, the agony ceases....",12,0,100,0,0,0,3681,0,"Servant of Razelikh"),
(7668,1,2,"Give my regards to Commander Trebor.",12,0,100,0,0,0,3682,0,"Servant of Razelikh"),
(7668,1,3,"The Defiler... must warn Thrall...",12,0,100,0,0,0,3683,0,"Servant of Razelikh"),
(7668,1,4,"It was not his fault. Tell him that...",12,0,100,0,0,0,3684,0,"Servant of Razelikh"),
(7668,1,5,"Finally, released from his cursed grasp. Father, I come... ",12,0,100,0,0,0,3685,0,"Servant of Razelikh");

-- Stone of Binding
UPDATE `gameobject_template` SET `AIName`="SmartGameObjectAI" WHERE `entry`=141812;
DELETE FROM `smart_scripts` WHERE `entryorguid`=141812 AND `source_type`=1;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(141812,1,0,0,11,0,100,0,0,0,0,0,105,16,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stone of Binding - On Respawn - Add Gameobject Flag 'Not Selectable'"),
(141812,1,1,0,70,0,100,0,2,0,0,0,11,10805,0,0,0,0,0,19,7668,50,0,0,0,0,0,"Stone of Binding - On Gameobject State Changed - Cast 'Shackle Shatter'");

-- Spell "Shackle Shatter"
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=13 AND `SourceEntry` IN (10805,10834,10835,10836);
INSERT INTO `conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry`, `SourceId`, `ElseGroup`, `ConditionTypeOrReference`, `ConditionTarget`, `ConditionValue1`, `ConditionValue2`, `ConditionValue3`, `NegativeCondition`, `ErrorType`, `ErrorTextId`, `ScriptName`, `Comment`) VALUES
(13,3,10805,0,0,31,0,3,7668,0,0,0,0,"","Spell 'Shackle Shatter' can only hit Servant of Razelikh"),
(13,3,10834,0,1,31,0,3,7669,0,0,0,0,"","Spell 'Shackle Shatter' can only hit Servant of Grol"),
(13,3,10835,0,2,31,0,3,7670,0,0,0,0,"","Spell 'Shackle Shatter' can only hit Servant of Allistarj"),
(13,3,10836,0,2,31,0,3,7671,0,0,0,0,"","Spell 'Shackle Shatter' can only hit Servant of Sevine");
```

**Steps to reproduce the problem:**

In the sql above the gob 141812 will cast the spell 10805 on the npc 7668, on spell_hit the npc will say something and dies

1. .go game 42469
2. .quest add 2681
3. click the gob
4. Notice the spell hits Servant of Razelikh but nothing happens
5. .die and .resp the npc and cast 10805 on it, in this case the event will be triggered as the caster is a player and not a gob

**Branch(es):** 3.3.5

**TC rev. hash/commit:**  rev. 7c003c568cdd

### PR DESCRIPTION
**Pull request by ForesterDev**
**Changes proposed:**
Core/SAI: fix SMART_EVENT_SPELLHIT with gobs as caster

**Target branch(es):** 3.3.5

**Issues addressed:** Closes #22613  

**Tests performed:** (Does it build, tested in-game, etc.) Tested

### CODE REVIEW

File: src/server/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -2328,7 +2328,26 @@ void Spell::TargetInfo::DoTargetSpellHit(Spell* spell, uint8 effIndex)
         return;                                             // No missinfo in that case
 
     if (_spellHitTarget)
+    {
```
Comment (Shauren) [inline src/server/game/Spells/Spell.cpp]: why did you change the order here?

Comment (ForesterDev) [reply src/server/game/Spells/Spell.cpp]: Because function  

void Spell::TargetInfo::DoDamageAndTriggers(Spell* spell)

rreturn in LINE when spell casted bu GO

// Get original caster (if exist) and calculate damage/healing from him data
     // Skip if m_originalCaster not available
     Unit* caster = spell->m_originalCaster ? spell->m_originalCaster : spell->m_caster->ToUnit();
     if (!caster)
         return;

So AI functions for gobj casts not called

Comment (Shauren) [reply src/server/game/Spells/Spell.cpp]: No, this is a bad change, `Spell::TargetInfo::DoTargetSpellHit` is called for each effect on target but you only want to call AI hooks once

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by Killyana** [comment]
@ForesterDev you will not finish this PR?

**Comment by illfated** [comment]
I presume that ForesterDev doesn't see how to fix the problem the way it should be solved (as outlined by Shauren), hence closing this.
